src/mbgl/map/annotation.cpp

```diff
@@ -300,13 +533,89 @@ LatLngBounds AnnotationManager::getBoundsForAnnotations(const AnnotationIDs& ids
+        ).first->second.second.get();
     }
-    return nullptr;
+
+    if (renderTile != nullptr && orderedShapeAnnotations.size()) {
+
+        // create shape tile layers from GeoJSONVT queries
+        for (auto& tiler_it : shapeTilers) {
+            const auto annotationID = tiler_it.first;
+            const std::string layerID = ShapeLayerID + "." + std::to_string(annotationID);
+
+            // check for existing render layer
+            auto renderLayer = renderTile->getMutableLayer(layerID);
+
+            if (renderLayer == nullptr) {
+                // we might need to create a tile layer for this shape
+                const auto& shapeTile = tiler_it.second->getTile(id.z, id.x, id.y);
+
+                if (shapeTile) {
+
+                    // shape exists on this tile; let's make a layer
+                    renderLayer = std::make_shared<LiveTileLayer>();
+
+                    // convert the features and add to render layer
+                    for (auto& shapeFeature : shapeTile.features) {
+
+                        using namespace mapbox::util::geojsonvt;
+
+                        FeatureType renderType = FeatureType::Unknown;
+
+                        if (shapeFeature.type == TileFeatureType::LineString) {
+                            renderType = FeatureType::LineString;
+                        } else if (shapeFeature.type == TileFeatureType::Polygon) {
+                            renderType = FeatureType::Polygon;
+                        }
+
+                        assert(renderType != FeatureType::Unknown);
+
+                        GeometryCollection renderGeometry;
+
+                        for (auto& shapeGeometry : shapeFeature.geometry) {
+
+                            std::vector<Coordinate> renderLine;
+
+                            auto& shapeRing = shapeGeometry.get<TileRing>();
+
+                            for (auto& shapePoint : shapeRing.points) {
+                                renderLine.emplace_back(shapePoint.x, shapePoint.y);
+                            }
+
+                            renderGeometry.push_back(renderLine);
+                        }
+
+                        auto renderFeature = std::make_shared<LiveTileFeature>(renderType, renderGeometry);
+
+                        renderLayer->addFeature(renderFeature);
+                    }
+
+                    // move the layer to the render tile
+                    renderTile->addLayer(layerID, renderLayer);
+
+                    // associate the annotation with the tile
+                    auto tile_update_it = tiles.find(id);
+                    assert(tile_update_it != tiles.end());
+                    tile_update_it->second.first.insert(annotationID);
+                }
+            }
+        }
+    }
+
+    return renderTile;
 }
 
-const std::string AnnotationManager::layerID = "com.mapbox.annotations.points";
+const std::string AnnotationManager::PointLayerID = "com.mapbox.annotations.points";
```

Comment:
And also, the idea was to namespace them since they are actual style layers, so as to not conflict with any named layers the user may have put in the style.